storage/fanout.go

```diff
@@ -231,22 +232,38 @@ func NewMergeQuerier(primaryQuerier Querier, queriers []Querier) Querier {
 func (q *mergeQuerier) Select(params *SelectParams, matchers ...*labels.Matcher) (SeriesSet, Warnings, error) {
 	seriesSets := make([]SeriesSet, 0, len(q.queriers))
 	var warnings Warnings
-	for _, querier := range q.queriers {
-		set, wrn, err := querier.Select(params, matchers...)
-		q.setQuerierMap[set] = querier
-		if wrn != nil {
-			warnings = append(warnings, wrn...)
-		}
-		if err != nil {
-			q.failedQueriers[querier] = struct{}{}
-			// If the error source isn't the primary querier, return the error as a warning and continue.
-			if querier != q.primaryQuerier {
-				warnings = append(warnings, err)
-				continue
-			} else {
-				return nil, nil, err
+	wg := sync.WaitGroup{}
+	wg.Add(len(q.queriers))
+	queryResult := make(map[int]SeriesSet)
+	var priErr error
+	isPriError := false
+	for idx, qr := range q.queriers {
+		go func(index int, querier Querier) {
+			defer wg.Done()
+			set, wrn, err := querier.Select(params, matchers...)
+			q.setQuerierMap[set] = querier
```

Comment:
You need locking on all these.

Reply:
Thank you for your review. According to my understanding, only 'q.setqueriermap' has the risk of concurrent modification, so the lock is only added to 'q.setqueriermap'

Lock.Lock ()
q.setQuerierMap[set] = querier
lock.Unlock()

Reply:
It's not the only one.

Reply:
![image](https://user-images.githubusercontent.com/9583245/73379486-47623900-42fd-11ea-87fd-86771643bf72.png)
I checked the go syntax. Waitgroup.done is not thread safe either. Two locks were added specially to ensure that the slowest query is not affected by the lock performance

Reply:
Huh? Waitgroup is threadsafe.